### PR TITLE
Add listing of lints (eg via `-W help`) to rustdoc

### LINKED ISSUE #83883: Rustdoc ignores -W help
I tried this code: `rustdoc -W help`
I expected to see this happen: Rustdoc prints a list of all lints, the same way rustc does.
Instead, this happened: Rustdoc ignored the flag.

### Meta

`rustdoc --version`: rustdoc 1.53.0-nightly (0b417ab5c 2021-04-03)

### PR DESCRIPTION
**Pull request by eggyal**
Fixes #83883

r? @jyn514

### CODE REVIEW

File: src/librustdoc/lib.rs [hunk 1]
```diff
@@ -651,6 +651,22 @@ fn main_options(options: config::Options) -> MainResult {
...
+                    register_lints(compiler.session(), &mut lint_store);
+                    true
+                } else {
+                    false
+                };
+            describe_lints(compiler.session(), &lint_store, registered_lints);
+            return Ok(());
+        }
         compiler.enter(|queries| {
             let sess = compiler.session();
```
Comment (jyn514) [inline src/librustdoc/lib.rs:671]: This is more complicated than it needs to be (it looks like you copied it from `rustc_driver`, which runs this before it registers lints).
```suggestion
        let sess = compiler.session();
        if sess.opts.describe_lints {
            let (_, lint_store) = &*compiler.register_plugins()?.peek();
            describe_lints(sess, &lint_store, true);
            return Ok(());
        }
        compiler.enter(|queries| {
```

File: src/test/rustdoc-ui/issue-83883-describe-lints.rs [hunk 2]
```diff
@@ -0,0 +1,2 @@
+// compile-flags: -W help
+// build-pass
```
Comment (jyn514) [inline src/test/rustdoc-ui/issue-83883-describe-lints.rs:2]:
```suggestion
// check-pass
```

**Review by jyn514 — COMMENTED** [review]
Thanks!

File: src/test/rustdoc-ui/issue-83883-describe-lints.rs (hunk 2 above)
Comment (eggyal) [reply src/test/rustdoc-ui/issue-83883-describe-lints.rs:2]: I did try with `check-pass` but think no output was produced... didn't investigate further, but let me check again...

File: src/librustdoc/lib.rs (hunk 1 above)
Comment (eggyal) [reply src/librustdoc/lib.rs]: Yes, you're right—I did just pillage from `rustc_driver`.  This is much better!  Thanks

File: src/test/rustdoc-ui/issue-83883-describe-lints.rs (hunk 2 above)
Comment (eggyal) [reply src/test/rustdoc-ui/issue-83883-describe-lints.rs]: Nope, all good!

**Comment by jyn514** [comment]
Looks like CI is stuck, but r=me when it's working.

**Comment by jyn514** [comment]
@bors r+

This awesome,  thank you! Nice job on normalizing the lint names :)

**Comment by camelid** [comment]
Hmm, won't it be confusing since a lot of rustc lints won't be run by rustdoc? E.g.,

```rust
fn main() {
    [1, 2, 3].into_iter();
}
```

with `rustdoc -W array-into-iter foo.rs` prints nothing, while a warning is printed when you run it with `rustc`.

**Comment by jyn514** [comment]
@camelid you can still enable those lints with `-W lint`; if you do that for an unknown lint it gives a hard error:
```
error[E0602]: unknown lint: `x`
  |
  = note: requested on the command line with `-W x`
```
I think the list of lints enabled by default is separate, maybe we could just document that somewhere.

**Comment by jyn514** [comment]
> with rustdoc -W array-into-iter foo.rs prints nothing, while a warning is printed when you run it with rustc.

Also, this is an implementation detail of how https://doc.rust-lang.org/nightly/rustdoc/advanced-features.html#cfgdoc-documenting-platform-specific-or-feature-specific-information works, it's not guarenteed.

**Comment by camelid** [comment]
> I think the list of lints enabled by default is separate, maybe we could just document that somewhere.

Yes, but I think it could be confusing that `-W array-into-iter` silently does nothing.

**Comment by jyn514** [comment]
> Yes, but I think it could be confusing that -W array-into-iter silently does nothing.

**Pull request merged**